src/banman.cpp

```diff
@@ -39,18 +40,23 @@ BanMan::~BanMan()
 
 void BanMan::DumpBanlist()
 {
-    SweepBanned(); // clean unused entries (if bantime has expired)
-
-    if (!BannedSetIsDirty()) return;
-
-    int64_t n_start = GetTimeMillis();
+    static Mutex dump_mutex;
+    LOCK(dump_mutex);
 
     banmap_t banmap;
-    GetBanned(banmap);
-    if (m_ban_db.Write(banmap)) {
+    {
+        LOCK(m_cs_banned);
+        SweepBanned();
+        if (!BannedSetIsDirty()) return;
+        banmap = m_banned;
         SetBannedSetDirty(false);
     }
 
+    int64_t n_start = GetTimeMillis();
+    if (!m_ban_db.Write(banmap)) {
+        SetBannedSetDirty(true);
```

Comment:
unrelated: Is m_dirty even used on shutdown when true? Shouldn't the below log be adjusted when it fails?